Document `Generator` based random number generation (#10134)
dask/dask: docs/source/array-random.rst

Random Number Generation
========================

Dask's random number routines produce pseudo random numbers using a backend
library via combinations of a ``BitGenerator`` to create sequences and a 
``Generator`` to use those sequences to sample from different statistical

> distributions:

Reviewer: This feels slightly confusing, because it's not clear how the "backend" library comes into play. Maybe leave that point out and clarify that point below? E.g:
 
"Dask's random number routines produce pseudo random numbers using combinations of a ``BitGenerator`` to create sequences and a ``Generator`` to use those sequences to sample from different statistical distributions:"

Or

"Dask produces pseudo-random numbers using a ``BitGenerator`` object to create random bit sequences and a ``Generator`` object to transform those bits into random values from a statistical distribution."

Author: Aye. It was a confusing first sentence. Went with the first version. Thank you

>     BitGenerators: Objects that generate random numbers. These are typically

Reviewer: Also, maybe bolden the "BitGenerators:"?

    unsigned integer words filled with sequences of either 32 or 64 random
    bits.

    Generators: Objects that transform sequences of random bits from a

>     BitGenerator into sequences of numbers that follow a specific probability

Reviewer: ```suggestion
    ``BitGenerator`` into sequences of numbers that follow a specific probability
```

    distribution (such as uniform, Normal or Binomial) within a specified
    interval.

Since Dask version 2023.2.1 the ``Generator`` can be initialized with a number

> of different ``BitGenerators``. It exposes many different probability

Reviewer: ```suggestion
of different ``BitGenerator`` classes. It exposes many different probability
```

distributions. The legacy ``RandomState`` random number routines are still
available, but are considered frozen and will not be getting any updates.

Differences with NumPy
----------------------

> Dask follows NumPy interface for random number generation with some differences:

Reviewer: ```suggestion
Dask follows the NumPy interface for random-number generation with some differences:
```

> - Methods under ``dask.array.random`` take ``chunks`` keyword.

Reviewer: ```suggestion
- Methods under ``dask.array.random`` take a ``chunks`` keyword.
```

- Dask tries to be backend agnostic. In other words, you can mostly use CuPy
  and NumPy interchangeably as a backend for random number generation. Any

>   library providing a similar interface should also work with minimal effort.

Author: haha true. that was naive :) Changed to `some effort`

Notes
-----

- Dask does not guarantee that the same number generator is used across versions.
  This means that numbers generated by ``dask.array.random`` by a new version may
  not be the same as the previous one, even when the same seed and distribution
  are used. As better algorithms evolve, the bit stream may change.
- Dask does not guarantee parity in the generated numbers with any third party
  library. In particular, numbers generated by Dask and NumPy or CuPy will differ

>   even when given the same seed and BitGenerator. Dask tends to spawn ``SeedSequence``

Reviewer: ```suggestion
  even when given the same seed and ``BitGenerator``. Dask tends to spawn ``SeedSequence``
```
